# Лабораторная работа №1

_Цель работы_ - изучить постановку антагонистической игры двух лиц в нормальной форме; найти решение игры за обоих игроков в смешанных стратегиях (стратегическую седловую точку).

## Постановка задачи и методические указания

Для игры, заданной матрицей $c_{ij}$, требуется найти оптимальные смешанные стратегии обоих игроков, сведя матричную игру к задаче ЛП (прямой для одного игрока и двойственной для другого).

Задачи ЛП следует решать симплекс-методом, приводя начальные, промежуточные и конечные симплекс-таблицы. По окончании алгоритма полученные решения необходимо проверить на допустимость.

### Вариант 5

## Ход работы

In [1]:
from IPython.core.display import display, HTML, Latex
latex.add_to_preamble('\usepackage[russian]{babel}')
pretty_print_default(True)

Зададим матрицу стратегий:

In [2]:
C = matrix(SR, 4, 5, [8, 1, 17, 8, 1, 12, 6, 11, 10, 16, 4, 19, 11, 15, 2, 17, 19, 6, 17, 16])
C

[ 8  1 17  8  1]
[12  6 11 10 16]
[ 4 19 11 15  2]
[17 19  6 17 16]

Сформулируем задачу линейного программирования для решения симплекс-методом для игрока $A$:

In [3]:
A = C.transpose()
b = [1] * A.nrows()
c = [1] * A.ncols()

P = InteractiveLPProblem(A, b, c, constraint_type=[">="] * len(b), problem_type="min", variable_type='>=')
P

LP problem (use typeset mode to see details)

Решим задачу ЛП симплекс-методом:

In [4]:
P = P.standard_form()
Latex(P.run_simplex_method())

<IPython.core.display.Latex object>

In [5]:
D = P.final_dictionary()

W = -D.objective_value()
g = 1/W

solution = D.basic_solution()

x = [i * g for i in solution]
x

[553/3684, 1555/3684, 773/3684, 803/3684]

Таким образом, оптимальная смешанная стратегия игрока $A$: $\left(\frac{553}{3684}, \frac{1555}{3684}, \frac{773}{3684}, \frac{803}{3684}\right)$.

Сформулируем задачу ЛП для игрока $B$:

In [6]:
B = C
b = [1] * B.nrows()
c = [1] * B.ncols()

P = InteractiveLPProblem(B, b, c, constraint_type=["<="] * len(b), problem_type="max", variable_type='>=')
P

LP problem (use typeset mode to see details)

In [7]:
P = P.standard_form()
Latex(P.run_simplex_method())

<IPython.core.display.Latex object>

In [8]:
D = P.final_dictionary()

Z = D.objective_value()
h = 1/Z

solution = D.basic_solution()

v = [i * h for i in solution]
v

[233/3684, 185/921, 719/1228, 0, 277/1842]

Таким образом, оптимальная смешанная стратегия игрока $B$: $\left(\frac{233}{3684}, \frac{185}{921}, \frac{719}{1228}, 0, \frac{277}{1842}\right)$.

## Контрольные вопросы

1. Определите понятие матричной игры с нулевой суммой.

Игры с нулевой суммой - антагонистические игры. В таком случае выигрыш одного игрока является проигрышем другого.

2. Дайте определение верхней и нижней цен игры. Докажите теорему о минимаксе.

Минимальный гарантированный выигрыш игрока $A$ называется *нижней ценой игры* и равняется $\max_{i}\min_{j}c_{ij}$. 

Минимальный возможный проигрыш игрока $B$ называется *верхней ценой игры*  и равняется $\min_{j}\max_{i} c_{ij}$.

3. Что такое цена игры? Сформулируйте теорему о седловой точке.

Если верхняя и нижняя цены игры равны, то это их значение назыается *ценой игры*.

**Теорема о седловой точке.** Для $(c_{ij})$ - произвольной матрицы $m \times n$ $$\max_{i \in A}\min_{j \in B} c_{ij} = \min_{j \in B}\max_{i \in A} c_{ij} = \min_{j \in B}\max_{i \in A}c_{ij}, $$ где $A = \overline{1, m}$, $B = \overline{1, n}$, тогда и только тогда, когда $(c_{ij})$ имеет седловую точку $(i_{0}, j_{0})$, для которой $c_{i_{0}j_{0}}$ является одновременно минимальным элементом строки и максимальным элементом столбца, и $\max_{i \in A}\min_{j \in B} c_{ij} = \min_{j \in B}\max_{i \in A} c_{ij} = c_{i_{0}j_{0}}$ - цена игры.

4. Сформулируйте основную теорему прямоугольных игр.

**Основная теорема прямоугольный игр (теорема Неймана).** Пусть задана матрица стратегий и выбраны стратегии $X = (x_{1}, \dots, x_{m}) \in S_{m}, Y = (y_{n}, \dots, y_{n}) \in S_{n}$; математическое ожидание выигрыша игрока $A$ имеет вид $$E(X, Y) = \sum_{i = 1}^{m}\sum_{j = 1}^{n} c_{ij}x_{i}y_{j};$$ тогда $$\max_{X \in S_{m}}\min_{Y \in S_{n}} E(X, Y) = \min_{Y \in S_{n}}\max_{X \in S_{m}} E(X, Y) = E(X^{*}, Y^{*}),$$ где $(X^{*}, Y^{*})$ - стратегическая седловая точка.

5. Что такое смешанные стратегии?

Если игра не имеет седловой точки, то на практике конкурирующие игроки применяют *смешанные* стратегии, т.е. попеременно используют две и более стратегии.

*Смешанная стратегия* игрока $A$ - это упорядоченная система $m$ действительных неотрицательных чисел $x_{i}, i = 1, \dots, m$ такая, что $\sum_{i = 1}^{m} x_{i} = 1$. Аналогично определяется стратегия игрока $B$ - $y_{j}, j = 1, \dots, m$: $\sum_{j = 1}^{n} y_{j} = 1$.